segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/MemoryStateUpdater.java

```diff
@@ -181,10 +253,40 @@ private void addToReadIndex(StorageOperation operation) {
         // Record recent activity on stream segment, if applicable.
         // We should record this for any kind of StorageOperation. When we issue 'triggerFutureReads' on the readIndex,
         // it should include 'sealed' StreamSegments too - any Future Reads waiting on that Offset will be cancelled.
-        synchronized (this.readIndex) {
-            this.recentStreamSegmentIds.add(operation.getStreamSegmentId());
+        return operation.getStreamSegmentId();
+    }
+
+    //endregion
+
+    //region CommitGroup
+
+    @NotThreadSafe
```

Comment:
It is hard to verify that `add()` will always be called in the same thread as `commit()` or that they will use the same external synchronization. As they are not invoked from OperationsProcessor which is not the outer class. Perhaps it would be easiest to just make it threadsafe.